sdk/containerregistry/container-registry/src/blob/containerRegistryBlobClient.ts

```diff
@@ -0,0 +1,400 @@
+      throw e;
+    } finally {
+      span.end();
+    }
+  }
+
+  /**
+   * Delete a manifest. Doing so effectively deletes an artifact from the registry.
+   *
+   * @param digest - the digest of the manifest to delete
+   * @param options - optional configuration used to send requests to the service
+   */
+  public async deleteManifest(digest: string, options?: DeleteManifestOptions): Promise<void> {
+    const { span, updatedOptions } = createSpan(
+      "ContainerRegistryBlobClient-deleteManifest",
+      options
+    );
+
+    try {
+      await this.client.containerRegistry.deleteManifest(
+        this.repositoryName,
+        digest,
+        updatedOptions
+      );
+    } catch (e) {
+      span.setStatus({ code: SpanStatusCode.ERROR, message: e.message });
+      throw e;
+    } finally {
+      span.end();
+    }
+  }
+
+  /**
+   * Upload an artifact blob.
+   *
+   * @param blobStreamFactory - a factory which produces a stream containing the blob data. This function may be called multiple times; each time the function is called a fresh stream should be returned.
+
+   */
+  public async uploadBlob(
+    blobStreamFactory: () => NodeJS.ReadableStream
+  ): Promise<UploadBlobResult>;
+
+  /**
+   * Upload an artifact blob.
+   *
+   * @param blobStream - the stream containing the blob data.
+   */
+  public async uploadBlob(blobStream: NodeJS.ReadableStream): Promise<UploadBlobResult>;
+
+  public async uploadBlob(
+    blobStreamOrFactory: (() => NodeJS.ReadableStream) | NodeJS.ReadableStream
+  ): Promise<UploadBlobResult> {
+    const { span } = createSpan("ContainerRegistryBlobClient-uploadBlob", undefined);
+
+    try {
+      const startUploadResult = await this.client.containerRegistryBlob.startUpload(
+        this.repositoryName
+      );
+
+      let requestBody: NodeJS.ReadableStream | Buffer;
+      let digest: string;
+
+      if (typeof blobStreamOrFactory === "function") {
+        requestBody = blobStreamOrFactory();
+        digest = await calculateDigest(blobStreamOrFactory());
+      } else {
+        requestBody = await readStreamToEnd(blobStreamOrFactory);
+        digest = await calculateDigest(requestBody);
+      }
+
+      if (!startUploadResult.location) {
+        throw new Error("startUpload did not return a location");
+      }
+
+      const uploadChunkResult = await this.client.containerRegistryBlob.uploadChunk(
+        startUploadResult.location.substring(1),
+        requestBody
+      );
+
+      if (!uploadChunkResult.location) {
+        throw new Error("uploadChunk did not return a location");
+      }
+
+      const { dockerContentDigest: digestFromResponse } =
+        await this.client.containerRegistryBlob.completeUpload(
+          digest,
+          uploadChunkResult.location.substring(1)
+        );
+
+      if (!digestFromResponse) {
+        throw new Error("completeUpload did not provide a digest");
+      }
+
+      if (digest !== digestFromResponse) {
```

Comment:
I wonder if we should make a custom DigestMismatchError or something, since this feels sort of data integrity/security related